src/synth/fluid_chan.c

```diff
@@ -294,3 +314,360 @@ fluid_channel_get_sfont_bank_prog(fluid_channel_t* chan, int *sfont,
+ */
+void fluid_channel_clear_monolist(fluid_channel_t* chan)
+{
+	/* keeps trace off the most recent note played */
+	chan->prev_note= chan->monolist[chan->i_last].note;
+
+	/* flush the monolist */
+	chan->i_first = chan->monolist[chan->i_last].next;
+	chan->n_notes = 0;
+	/* Update legato/ sataccato playing state */
+	chan->mode &= ~ FLUID_CHANNEL_LEGATO_PLAYING; /* Staccato state */
+}
+
+/**
+ * The monophonic list is flushed keeping last note only.
+ * The function is entended to be called on legato Off when the
+ * monolist have notes.
+ *
+ *            The monophonic list
+ *  +------------------------------------------------+
+ *  | +--------------------------------------------+ |
+ *  | |  +----+   +----+          +----+   +----+  | |
+ *  | +--|note|<--|note|<--....<--|note|<--|note|<-+ |
+ *  +--->|vel |-->|vel |-->....-->|vel |-->|vel |----+
+ *       +----+   +----+          +----+   +----+
+ *        /|\                      /|\
+ *         |                        |
+ *      i_first                   i_last
+ *
+ * @param chan  fluid_channel_t.
+ * Note: i_last index keeps a trace of the most recent note played.
+ *       prev_note keeps a trace of the note .
+ *       FLUID_CHANNEL_LEGATO_PLAYING bit keeps trace of legato/staccato playing.
+ */
+void fluid_channel_keep_lastnote_monolist(fluid_channel_t* chan)
+{
+	chan->i_first = chan->i_last;
+	chan->n_notes = 1;
+}
+
+/**
+ * On noteOn on a polyphonic channel,adds the note into the monophonic list
+ * keeping only this note.
+ * @param 
+ *   chan  fluid_channel_t.
+ *   key, vel, note and velocity added in the monolist
+ * Note: i_last index keeps a trace of the most recent note inserted.
+ *       prev_note keeps a trace of the note prior i_last note.
+ *       FLUID_CHANNEL_LEGATO_PLAYING bit keeps trace of legato/staccato playing.
+ */
+void fluid_channel_set_onenote_monolist(fluid_channel_t* chan, unsigned char key,
+								            unsigned char vel)
+{
+	fluid_channel_add_monolist(chan, key, vel,1);
+}
+
+/**
+ * The function changes the state (Valid/Invalid) of the previous note played in
+ * a staccato manner (fluid_channel_prev_note()).
+ * When potamento mode: 'each note' or 'staccato only' is selected, on next 
+ * noteOn a portamento will be started from the most recent note played 
+ * staccato fluid_channel_last_note.
+ * It will be possible that it isn't appropriate. To give the musician the 
+ * possibility to choose this note , prev_note will be marked invalid on noteOff
+ * if portamento pedal is Off.
+ *
+ * The function is intended to be called when the following event occurs:
+ * - On noteOff (in poly or mono mode), to mark prev_note invalid.
+ * - On Portamento Off(in poly or mono mode), to mark prev_note invalid.
+ * @param chan  fluid_channel_t.
+ */
+void fluid_channel_invalid_prev_note_staccato(fluid_channel_t* chan)
+{
+	if(!(chan->mode  & FLUID_CHANNEL_LEGATO_PLAYING)) /* the monophonic list is empty */ 
```

Comment:
If you want to check whether the list is empty, couldnt you explicitly check `chan->n_notes == 0`?

Reply:
Here the comment  `/* the monophonic list is empty */` is in incorrect and should be `/* check if the playing is staccato */`. This correction will be done.